src/invariant/test/OrderBookIsNotCrossedTests.cpp

```diff
@@ -0,0 +1,274 @@
+        REQUIRE(orders.at(offer1.data.offer().offerID).price.d == 2);
+
+        ltxPtr->commit();
+
+        SECTION("Then modify offer")
+        {
+            auto offer2 = generateOffer(cur1, cur2, 2, Price{5, 3});
+            offer2.data.offer().sellerID = offer1.data.offer().sellerID;
+            offer2.data.offer().offerID = offer1.data.offer().offerID;
+
+            auto ltxStore =
+                std::make_unique<LedgerTxn>(app->getLedgerTxnRoot());
+            auto ltxPtr = ltxStore.get();
+            auto entry = ltxPtr->load(LedgerEntryKey(offer1));
+            entry.current() = offer2;
+
+            invariant->checkOnOperationApply(op, opRes, ltxPtr->getDelta());
+            auto orderbook = invariant->getOrderBook();
+            auto orders = orderbook[cur1AssetId][cur2AssetId];
+
+            REQUIRE(orders.size() == 1);
+            REQUIRE(orders.at(offer2.data.offer().offerID).amount == 2);
+            REQUIRE(orders.at(offer2.data.offer().offerID).price.n == 5);
+            REQUIRE(orders.at(offer2.data.offer().offerID).price.d == 3);
+        }
+
+        SECTION("Then delete offer")
+        {
+            auto ltxStore =
+                std::make_unique<LedgerTxn>(app->getLedgerTxnRoot());
+            auto ltxPtr = ltxStore.get();
+            auto entry = ltxPtr->load(LedgerEntryKey(offer1));
+            entry.erase();
+
+            invariant->checkOnOperationApply(op, opRes, ltxPtr->getDelta());
+            auto orderbook = invariant->getOrderBook();
+            auto orders = orderbook[cur1AssetId][cur2AssetId];
+
+            REQUIRE(orders.size() == 0);
+        }
+    }
+}
+
+TEST_CASE("OrderBookIsNotCrossed properly throws if order book is crossed",
+          "[invariant][OrderBookIsNotCrossed]")
+{
+    Config cfg = getTestConfig(0);
+    cfg.INVARIANT_CHECKS = {"OrderBookIsNotCrossed"};
+
+    VirtualClock clock;
+    Application::pointer app = createTestApplication(clock, cfg);
+
+    Asset cur1;
+    cur1.type(ASSET_TYPE_CREDIT_ALPHANUM4);
+    strToAssetCode(cur1.alphaNum4().assetCode, "CUR1");
+
+    Asset cur2;
+    cur2.type(ASSET_TYPE_CREDIT_ALPHANUM4);
+    strToAssetCode(cur2.alphaNum4().assetCode, "CUR2");
+
+    // the initial state for the order book is:
+    //     3 A @  3/2 (B/A)
+    //     2 A @  5/3 (B/A)
+    //     4 B @  3/4 (A/B)
+    //     1 B @  4/5 (A/B)
+    // where A = cur1 and B = cur2
+    genApplyCheckCreateOffer(cur1, cur2, 3, Price{3, 2}, *app, SHOULD_PASS);
+    genApplyCheckCreateOffer(cur1, cur2, 2, Price{5, 3}, *app, SHOULD_PASS);
+    genApplyCheckCreateOffer(cur2, cur1, 4, Price{3, 4}, *app, SHOULD_PASS);
+    genApplyCheckCreateOffer(cur2, cur1, 1, Price{4, 5}, *app, SHOULD_PASS);
+
+    SECTION("Not crossed when highest bid < lowest ask")
+    {
+        SECTION("Create offer")
+        {
+            // Offer 5: 7 A @  8/5 (B/A)
+            auto offer5 = generateOffer(cur1, cur2, 7, Price{8, 5});
```

Comment:
Why not use `genApplyCheckCreateOffer` here? (This question applies in multiple places in this file)